File: algorithms/Jump Game/Jump Game.js

var canJump = function(nums) {
    let target = nums.length-1;
    let max = 0,index = 0;
    
    while(index <= target){
        max = Math.max(max,index + nums[index]);
        
        if(max >= target) return true;
        
        if(index >= max && nums[index] === 0) return false;
        
        index++;
    }
    
    return false;
};

File: algorithms/J/Jump Game/Jump Game.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int ans = nums[0];
        if(nums.size()>1&&ans==0) return false;
        for(int i=1; i<nums.size(); i++){
            ans = max(ans-1,nums[i]);
            if(ans<=0&&i!=nums.size()-1) return false;
        }
        return true;
    }
};
